#8 Implemented Abstraction for Logging users

mparq opened the pull request.

Commit:
Created a file for storing LogManagers which give interfaces for logging in code. As of now, exposes only two methods writeToLog and flushLog. Two Managers were made: one for logging to notebook, and one for logging user choices. In the future, Manager for user choices might want to be separated out into its own abstraction.

Commit:
Created a file for storing LogManagers which give interfaces for logging in code. As of now, exposes only two methods writeToLog and flushLog. Two Managers were made: one for logging to notebook, and one for logging user choices. In the future, Manager for user choices might want to be separated out into its own abstraction.

Commit:
Created a file for storing LogManagers which give interfaces for logging in code. As of now, exposes only two methods writeToLog and flushLog. Two Managers were made: one for logging to notebook, and one for logging user choices. In the future, Manager for user choices might want to be separated out into its own abstraction.

Commit:
Modified PageController class to use the new Logging Abstractions.
	Now, logging through the api is handled through ChoiceLogManager.
	However, we make assumptions that we can use the aggregate requests api function in every situation
	Error handling may need to be added for logging (especially for api calls).
	New changePage method added which will take care of flushing logs and marking time spent on page before changing page for each button click.
	Moved logUserActions functionality to ChoiceLogManager.

Some notes on the design of the Logging Abstraction:
	I considered using a more specific LabNotebook logging abstraction,
	but in the end decided on a more general (user builds string which goes into log. log can then be flushed to notebook) approach.
	I don't know which way would be better, but I think this way is a little bit more flexible. The PageController class (which will eventually be changed)
	still needs to know how to build the Lab Notebook log string. If there is a better way, we should consider it.

Commit:
Modified PageController class to use the new Logging Abstractions.
	Now, logging through the api is handled through ChoiceLogManager.
	However, we make assumptions that we can use the aggregate requests api function in every situation
	Error handling may need to be added for logging (especially for api calls).
	New changePage method added which will take care of flushing logs and marking time spent on page before changing page for each button click.
	Moved logUserActions functionality to ChoiceLogManager.

Some notes on the design of the Logging Abstraction:
	I considered using a more specific LabNotebook logging abstraction,
	but in the end decided on a more general (user builds string which goes into log. log can then be flushed to notebook) approach.
	I don't know which way would be better, but I think this way is a little bit more flexible. The PageController class (which will eventually be changed)
	still needs to know how to build the Lab Notebook log string. If there is a better way, we should consider it.

Commit:
Modified PageController class to use the new Logging Abstractions.
	Now, logging through the api is handled through ChoiceLogManager.
	However, we make assumptions that we can use the aggregate requests api function in every situation
	Error handling may need to be added for logging (especially for api calls).
	New changePage method added which will take care of flushing logs and marking time spent on page before changing page for each button click.
	Moved logUserActions functionality to ChoiceLogManager.

Some notes on the design of the Logging Abstraction:
	I considered using a more specific LabNotebook logging abstraction,
	but in the end decided on a more general (user builds string which goes into log. log can then be flushed to notebook) approach.
	I don't know which way would be better, but I think this way is a little bit more flexible. The PageController class (which will eventually be changed)
	still needs to know how to build the Lab Notebook log string. If there is a better way, we should consider it.

calebdre committed:
got the engine working just like how pagecontroller was

calebdre committed:
fixed merge conflict

calebdre commented:
Can you make a new commit copying the functionality to the PatientEngine?

calebdre commented:
Also, make sure that the LogManagers are the ones that know how to assemble an action string.

calebdre committed:
added a new model and manager. cleaned up unused files

Commit:
Added new LogManager scripts to index.html

Commit:
Added new LogManager scripts to index.html

Commit:
Added new LogManager scripts to index.html

calebdre committed:
implemented patient tracking inside the patient Engine. Alsomade localstorage helper

calebdre committed:
implemented storagehelper

Commit:
Integrated logging changes into PatientEngine.
ChoiceLogManager now takes in a choiceInfo object and logs information about a choice
Action String now populated within ChoiceLogManager
LabNotebookLogManager tabled for now. Maybe we will want to make this more specific in the future.
Fixes made to PatientEngine to allow it to work.
	BUG: Popup windows don't behave correctly.

Commit:
Integrated logging changes into PatientEngine.
ChoiceLogManager now takes in a choiceInfo object and logs information about a choice
Action String now populated within ChoiceLogManager
LabNotebookLogManager tabled for now. Maybe we will want to make this more specific in the future.
Fixes made to PatientEngine to allow it to work.
	BUG: Popup windows don't behave correctly.

Commit:
Remove getActionString from PageController.js
. Change ChoiceLogManager to take in choices instead of messages.

mparq commented:
ChoiceLogManager should be good to go. Right now, it handles logging the users choice action string and sending that to the server. Do you want me to also make LabNotebookLogManager specific as well? Meaning, it will only handle writing those specific logs to the notebook within the simulation. Originally, I made it just a general wrapper into the LabNotebook so the user would still have to build their own string and pass that into the NotebookLogManager. I feel like we wouldn't even need this, since labnotebook provides its own add method. It also doesn't make sense to have a specific LabNotebookLogManager inside PatientEngine, because there are other places where we add to the lab notebook. Within PatientEngine, we are writing specific strings to the lab notebook, but maybe this functionality should be handled within our ChoiceLogManager as well. I can try to make this happen, but it does appear to be a little difficult due to the difference in building loggableString between onBinaryClick and onSubmitClick.

Commit:
Made ChoiceLogManager handle logging to notebook during simulation.

Commit:
Made ChoiceLogManager handle logging to notebook during simulation.

mparq commented:
Now ChoiceLogManager handles all logging during the simulation. However, some bugs have been introduced. From what I can tell, the code to grab the current Patient if it exists on the page is flawed. I'll have to go back to previous commits and see exactly where, but my suspicion is it happened in the move from PageController to PatientEngine. Another bug I noticed is that popup windows are no longer handled correctly. This may be due to something I changed. 

I also wanted to ask something, and this is in regards to the async api calls we make when logging the action string of a choice. Before, you had the code for changing the page attached as a callback for after the api call completes. I moved the code for changing code to not be attached as a call back since 1) ChoiceLogManager didn't seem like it should be handling page changes and 2) it made the code look a little bit nicer. Do you think this will introduce anything dangerous in the code? If so, we'll want to think of what we can do to make the page change wait until after the api call completes.

Commit:
Merged LogManager branch into new PatientEngine with StorageHelper.

Commit:
Quickfix for bug Object.values is not a function in line 30 of PatientEngine.

mparq commented:
Merged with latest upstream making use of LocalStorage and page contexts. We still have the problem of logging looking a little off regarding identifying When we should have a message that looks like "Patient:" and one that looks like "Question:". I also commented out the lines in PatientEngine which attempt to identify whether the page is a popup. I get an error in my browser stating Object.values is not a function. Does this work for your browser?

calebdre commented on scisim/static/js/Utils/LogManager.js:
```diff
@@ -0,0 +1,54 @@
+    labnotebook.add(loggableString.replace("}", ""));
+
+    if (choiceInfo.page_context.hasOwnProperty("patient")) {
+        choiceInfo.page_context.patient.choices.push(loggableString);
```
note that when you update a context, you have to call chain.updateContext on the context.

calebdre commented on scisim/static/js/Engines/Patient/PatientEngine.js:
```diff
@@ -1,11 +1,146 @@
+};
+
+PatientEngine.prototype.changePage = function(destination) {
+	this.choiceLogger.flushLog();
```
it doesn't look like the new page will be rendered before the api requests finish, however i don't think this is a problem. If it does become one, we can wrap it in a deferred.

Commit:
Updated chain context in logChoice. Modified to use storageHelper in flushLog.

calebdre commented:
LogManager looks pretty good - it cleans up the PatientEngine pretty nicely. I don't think the new page is rendered after the logging api calls are made, but i don't think this is an issue.  

For logging, we just want to append the patient name before the log if the name is available. I used the patient object in places however the initial detecting a patient doesn't seem to be working - i'll have to look into that later today. Right now, the logging doesn't have a bug but the PatientManager does.  

Object.Values should be working..i noticed you tried it before you pulled the latest version - have you tried it with the new changes? It works on my browser (Firefox.) I added it to the polyfils/Object because it's currently not a native Javascript method

mparq commented:
I did try it with the latest version I believe. Let me try it one more time to make sure. I also fixed up LogManager to call chain.updateContext and use storageHelper.

mparq commented:
Yeah, strange. I see the polyfils/Object.js and it is being imported in index.html, but I keep getting that type error Object.values is not a function. I tried in both Chrome and Firefox with the same result.

mparq commented:
Ah, okay. In my index.html, js/polyfils/Object.js was being imported, but it should have been js/Polyfils/Object.js. I guess I'll also commit that fix and uncomment the lines that were causing issue before. Hopefully this fixes the popup issue I was getting.

Commit:
Fixed typo in index.html causing Object.js not to be imported.

Commit:
Fixed bug in logChoice: choiceInfo.page_context instead of page_context.

Commit:
Fixed bug: reference to getJSON of storageHelper instead of getJson.

mparq commented:
Yep! Popups are work correctly now.
